web/js/codeworld.js

```diff
@@ -892,15 +894,58 @@ function formatSource() {
     return;
   }
 
-  const src = window.codeworldEditor.getValue();
+  const src = doc.getValue();
   const data = new FormData();
   data.append('source', src);
   data.append('mode', window.buildMode);
 
+  const oldCursorPosition = doc.getCursor();
+  const lineContentAtOldCursorPosition = doc.getLine(oldCursorPosition.line);
+
   sendHttp('POST', 'indent', data, (request) => {
     if (request.status === 200) {
-      if (request.responseText !== src) {
-        codeworldEditor.getDoc().setValue(request.responseText);
+      const reformattedSrc = request.responseText;
+
+      if (reformattedSrc !== src) {
+        doc.setValue(reformattedSrc);
+
+        let oldLine = 0;
+        let newLine = 0;
+
+        Diff.diffLines(src, reformattedSrc).forEach((diff) => {
+          const { value, count, added, removed } = diff;
+
+          if (added) {
+            if (oldCursorPosition.line <= oldLine) {
+              doc.setCursor(newLine, oldCursorPosition.ch);
+            }
+
+            const currentLine = doc.getCursor().line;
+            if (currentLine === newLine) {
+              const trimmedLineContent = lineContentAtOldCursorPosition.trimLeft();
+              const numberOfDeletedSpaces =
+                lineContentAtOldCursorPosition.length -
+                trimmedLineContent.length;
+              const newCharPosition =
+                value.indexOf(trimmedLineContent) +
+                oldCursorPosition.ch -
+                numberOfDeletedSpaces;
+
+              doc.setCursor(currentLine, newCharPosition);
+            }
+
+            newLine += count;
+          } else if (removed) {
+            oldLine += count;
+
+            if (oldCursorPosition.line >= oldLine) {
+              doc.setCursor(newLine, oldCursorPosition.ch);
+            }
+          } else {
+            newLine += count;
+            oldLine += count;
+          }
+        });
       }
```

Comment:
I think it will eventually become important to handle not just the cursor position, but the scroll position, as well.  That doesn't need to be in the first pull request, though.  This would definitely be an improvement on its own.

Reply:
What is it exactly about the scroll position that has to be handled?

Reply:
What I am concerned about is that it will be a jolting experience to run the formatter, and have the screen change scrolling, even if the cursor remains in the same place.  Ideally, we'd try to rescroll the editor so that the line with the cursor remains at the same offset from the top of the editor.  I think you could do this with something like:

```
const oldScroll = window.codeworldEditor.getScrollInfo();
const oldCursorPos = window.codeworldEditor.cursorCoords();

... set the new cursor position ...

const newCursorPos = window.codeworldEditor.cursorCoords();
window.codeworldEditor.scrollTo({
  x: oldScroll.left,
  y: oldScroll.top + newCursorPos.top - oldCursorPos.top
});
```

That intentionally doesn't change the horizontal scroll, because horizontal scrolling is unusual.